File: apps/desktop/src/lib/backend/memory.ts
import { requestJson, sendJson } from './client';

export type MemoryKind = 'stable' | 'episodic' | 'session_summary';
export type MemoryImportance = 'low' | 'medium' | 'high';
export type MemoryState = 'active' | 'deleted';
export type MemoryRecallStatus = 'active' | 'hidden';
export type MemoryStateFilter = 'active' | 'hidden' | 'deleted';
export type MemoryMode = 'all' | 'manual' | 'automatic';

export interface MemoryItemRecord {
  id: string;
  kind: MemoryKind;
  title: string;
  content: string;
  scope: string;
  source_kind: string;
  source_label: string;
  importance: MemoryImportance;
  state: MemoryState;
  recall_status: MemoryRecallStatus;
  is_manual: boolean;
  is_override: boolean;
  origin_session_id: string | null;
  origin_subagent_session_id: string | null;
  original_memory_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface MemoryItemsResponse {
  items: MemoryItemRecord[];
}

export interface MemoryHistoryEntryRecord {
  id: string;
  memory_id: string;
  action: string;
  summary: string | null;
  snapshot: Record<string, unknown> | null;
  created_at: string;
}

export interface MemoryHistoryResponse {
  items: MemoryHistoryEntryRecord[];
}

export interface MemoryRecallItemRecord {
  memory_id: string;
  title: string;
  kind: MemoryKind;
  scope: string;
  source_kind: string;
  source_label: string;
  importance: MemoryImportance;
  reason: string;
  origin_session_id: string | null;
  origin_subagent_session_id: string | null;
}

export interface MemoryRecallDetailRecord {
  assistant_message_id: string;
  session_id: string;
  created_at: string;
  reason_summary: string | null;
  items: MemoryRecallItemRecord[];
}

export interface MemoryRecallLogEntryRecord extends MemoryRecallDetailRecord {
  id: string;
  task_run_id: string | null;
}

export interface MemoryRecallLogResponse {
  items: MemoryRecallLogEntryRecord[];
}

export interface SessionRecallSummaryRecord {
  assistant_message_id: string;
  created_at: string;
  recalled_count: number;
  reason_summary: string | null;
  items: MemoryRecallItemRecord[];
}

export interface SessionRecallSummariesResponse {
  items: SessionRecallSummaryRecord[];
}

export interface MemoryItemCreateInput {
  kind: MemoryKind;
  title: string;
  content: string;
  scope: string;
  importance: MemoryImportance;
}

export interface MemoryItemUpdateInput {
  title?: string;
  content?: string;
  scope?: string;
  importance?: MemoryImportance;
}

export interface MemoryItemsQuery {
  kind?: MemoryKind;
  query?: string;
  scope?: string;
  sourceKind?: string;
  state?: MemoryStateFilter;
  recallStatus?: MemoryRecallStatus;
  mode?: MemoryMode;
}

function queryString(filters: MemoryItemsQuery = {}) {
  const params = new URLSearchParams();
  if (filters.kind) {
    params.set('kind', filters.kind);
  }
  if (filters.query) {
    params.set('query', filters.query);
  }
  if (filters.scope) {
    params.set('scope', filters.scope);
  }
  if (filters.sourceKind) {
    params.set('source_kind', filters.sourceKind);
  }
  if (filters.state) {
    params.set('state', filters.state);
  }
  if (filters.mode) {
    params.set('mode', filters.mode);
  }
  if (filters.recallStatus) {
    params.set('recall_status', filters.recallStatus);
  }
  const serialized = params.toString();
  return serialized ? `?${serialized}` : '';
}

export function fetchMemoryItems(
  filters: MemoryItemsQuery = {},
): Promise<MemoryItemsResponse> {
  return requestJson<MemoryItemsResponse>(`/memory/items${queryString(filters)}`, {
    method: 'GET',
  });
}

export function fetchMemoryItem(memoryId: string): Promise<MemoryItemRecord> {
  return requestJson<MemoryItemRecord>(`/memory/items/${memoryId}`, {
    method: 'GET',
  });
}

export function fetchMemoryItemHistory(
  memoryId: string,
): Promise<MemoryHistoryResponse> {
  return requestJson<MemoryHistoryResponse>(`/memory/items/${memoryId}/history`, {
    method: 'GET',
  });
}

export function createMemoryItem(
  payload: MemoryItemCreateInput,
): Promise<MemoryItemRecord> {
  return sendJson<MemoryItemRecord>('POST', '/memory/items', payload);
}

export function updateMemoryItem(
  memoryId: string,
  payload: MemoryItemUpdateInput,
): Promise<MemoryItemRecord> {
  return sendJson<MemoryItemRecord>('PUT', `/memory/items/${memoryId}`, payload);
}

export function hideMemoryItem(memoryId: string): Promise<MemoryItemRecord> {
  return sendJson<MemoryItemRecord>('POST', `/memory/items/${memoryId}/hide`);
}

export function restoreMemoryItem(memoryId: string): Promise<MemoryItemRecord> {
  return sendJson<MemoryItemRecord>('POST', `/memory/items/${memoryId}/restore`);
}

export function promoteMemoryItem(memoryId: string): Promise<MemoryItemRecord> {
  return sendJson<MemoryItemRecord>('POST', `/memory/items/${memoryId}/promote`);
}

export function demoteMemoryItem(memoryId: string): Promise<MemoryItemRecord> {
  return sendJson<MemoryItemRecord>('POST', `/memory/items/${memoryId}/demote`);
}

export function deleteMemoryItem(
  memoryId: string,
  hard = false,
): Promise<MemoryItemRecord | null> {
  return requestJson<MemoryItemRecord | null>(
    `/memory/items/${memoryId}${hard ? '?hard=true' : ''}`,
    {
      method: 'DELETE',
    },
  );
}

export function fetchMemoryRecallLog(): Promise<MemoryRecallLogResponse> {
  return requestJson<MemoryRecallLogResponse>('/memory/recall', {
    method: 'GET',
  });
}

export function fetchMemoryRecallDetail(
  messageId: string,
): Promise<MemoryRecallDetailRecord> {
  return requestJson<MemoryRecallDetailRecord>(
    `/memory/recall/messages/${messageId}`,
    { method: 'GET' },
  );
}

export function fetchSessionRecallSummaries(
  sessionId: string,
): Promise<SessionRecallSummariesResponse> {
  return requestJson<SessionRecallSummariesResponse>(
    `/memory/recall/sessions/${sessionId}`,
    { method: 'GET' },
  );
}

File: apps/backend/app/schemas/memory.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class MemoryEntryCreate(BaseModel):
    scope_type: str
    scope_key: str
    conversation_id: str | None = None
    session_id: str | None = None
    parent_session_id: str | None = None
    title: str = Field(min_length=1, max_length=200)
    body: str = Field(min_length=1, max_length=12000)
    summary: str | None = Field(default=None, max_length=4000)
    importance: float = Field(default=0.5, ge=0.0, le=1.0)
    confidence: float = Field(default=0.5, ge=0.0, le=1.0)
    expires_at: datetime | None = None

    @field_validator("scope_type", "scope_key", "title", "body", mode="before")
    @classmethod
    def _trim_required_strings(cls, value: str) -> str:
        return value.strip()

    @field_validator("summary", mode="before")
    @classmethod
    def _trim_optional_summary(cls, value: str | None) -> str | None:
        return value.strip() if isinstance(value, str) else value


class MemoryEntryUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    body: str | None = Field(default=None, min_length=1, max_length=12000)
    summary: str | None = Field(default=None, max_length=4000)
    importance: float | None = Field(default=None, ge=0.0, le=1.0)
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    expires_at: datetime | None = None

    @field_validator("title", "body", mode="before")
    @classmethod
    def _trim_optional_strings(cls, value: str | None) -> str | None:
        return value.strip() if isinstance(value, str) else value

    @field_validator("summary", mode="before")
    @classmethod
    def _trim_summary(cls, value: str | None) -> str | None:
        return value.strip() if isinstance(value, str) else value


class MemoryEntryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    agent_id: str | None = None
    scope_type: str
    scope_key: str
    conversation_id: str | None
    session_id: str | None
    root_session_id: str | None = None
    parent_session_id: str | None
    source_kind: str
    lifecycle_state: str
    title: str
    body: str
    summary: str | None
    importance: float
    confidence: float
    dedupe_hash: str
    created_at: datetime
    updated_at: datetime
    created_by: str
    updated_by: str
    workspace_path: str | None = None
    user_scope_key: str | None = None
    expires_at: datetime | None
    redaction_state: str
    security_state: str
    hidden_from_recall: bool
    deleted_at: datetime | None
    origin_message_id: str | None = None
    origin_task_run_id: str | None = None
    override_target_entry_id: str | None = None


class MemoryEntriesListResponse(BaseModel):
    items: list[MemoryEntryRead]


class MemoryHistoryItemRead(BaseModel):
    id: str
    memory_id: str
    action: str
    actor_type: str
    actor_id: str | None
    before_snapshot: dict[str, Any] | None
    after_snapshot: dict[str, Any] | None
    created_at: datetime


class MemoryEntryHistoryResponse(BaseModel):
    items: list[MemoryHistoryItemRead]


class MemoryDeleteResponse(BaseModel):
    deleted: bool


MemoryScopeName = Literal[
    "current_conversation",
    "current_session_tree",
    "agent",
    "user",
    "workspace",
]
MemorySearchRecordType = Literal["memory_entry", "session_summary"]
MemorySearchOverrideStatus = Literal["none", "overrides_automatic", "overridden_by_manual"]


class MemoryScopeContextRead(BaseModel):
    agent_id: str | None
    session_id: str | None
    root_session_id: str | None
    workspace_path: str | None
    user_scope_key: str | None


class MemoryScopeSupportRead(BaseModel):
    name: MemoryScopeName
    available: bool


class MemorySearchOriginRead(BaseModel):
    table: str
    agent_id: str | None
    session_id: str | None
    root_session_id: str | None
    origin_message_id: str | None
    origin_task_run_id: str | None
    workspace_path: str | None
    scope_type: str | None = None
    scope_key: str | None = None
    matched_scopes: list[MemoryScopeName] = Field(default_factory=list)


class MemorySearchOverrideRead(BaseModel):
    status: MemorySearchOverrideStatus
    target_id: str | None = None
    effective_id: str | None = None
    selected_via_substitution: bool = False


class MemorySearchItemRead(BaseModel):
    record_type: MemorySearchRecordType
    id: str
    title: str | None = None
    summary: str | None
    body: str | None
    source_kind: str
    importance: float
    score: float
    score_breakdown: dict[str, Any] = Field(default_factory=dict)
    origin: MemorySearchOriginRead
    override: MemorySearchOverrideRead


class MemorySearchResponse(BaseModel):
    query: str
    normalized_query: str
    applied_scopes: list[MemoryScopeName]
    context: MemoryScopeContextRead
    items: list[MemorySearchItemRead]


class MemoryRecallPreviewResponse(MemorySearchResponse):
    run_id: str | None = None


class MemoryScopesResponse(BaseModel):
    context: MemoryScopeContextRead
    default_scopes: list[MemoryScopeName]
    supported_scopes: list[MemoryScopeSupportRead]


MemoryKind = Literal["stable", "episodic", "session_summary"]
MemoryImportance = Literal["low", "medium", "high"]
MemoryState = Literal["active", "deleted"]
MemoryRecallStatus = Literal["active", "hidden"]
MemoryListStateFilter = Literal["active", "hidden", "deleted"]
MemoryMode = Literal["all", "manual", "automatic"]


class MemoryItemRead(BaseModel):
    id: str
    kind: MemoryKind
    title: str
    content: str
    scope: str
    source_kind: str
    source_label: str
    importance: MemoryImportance
    state: MemoryState
    recall_status: MemoryRecallStatus
    is_manual: bool
    is_override: bool
    origin_session_id: str | None = None
    origin_subagent_session_id: str | None = None
    original_memory_id: str | None = None
    created_at: datetime
    updated_at: datetime


class MemoryItemCreate(BaseModel):
    kind: MemoryKind
    title: str = Field(min_length=1, max_length=200)
    content: str = Field(min_length=1)
    scope: str = Field(default="global", min_length=1, max_length=100)
    importance: MemoryImportance = "medium"


class MemoryItemUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    content: str | None = Field(default=None, min_length=1)
    scope: str | None = Field(default=None, min_length=1, max_length=100)
    importance: MemoryImportance | None = None


class MemoryItemsResponse(BaseModel):
    items: list[MemoryItemRead]


class MemoryHistoryEntryRead(BaseModel):
    id: str
    memory_id: str
    action: str
    summary: str | None = None
    snapshot: dict[str, object] | None = None
    created_at: datetime


class MemoryItemHistoryResponse(BaseModel):
    items: list[MemoryHistoryEntryRead]


class MemoryRecallItemRead(BaseModel):
    memory_id: str
    title: str
    kind: MemoryKind
    scope: str
    source_kind: str
    source_label: str
    importance: MemoryImportance
    reason: str
    origin_session_id: str | None = None
    origin_subagent_session_id: str | None = None


class MemoryRecallDetailRead(BaseModel):
    assistant_message_id: str
    session_id: str
    created_at: datetime
    reason_summary: str | None = None
    items: list[MemoryRecallItemRead]


class MemoryRecallLogEntryRead(MemoryRecallDetailRead):
    id: str
    task_run_id: str | None = None


class MemoryRecallLogResponse(BaseModel):
    items: list[MemoryRecallLogEntryRead]


class SessionRecallSummaryRead(BaseModel):
    assistant_message_id: str
    created_at: datetime
    recalled_count: int
    reason_summary: str | None = None
    items: list[MemoryRecallItemRead]


class SessionRecallSummariesResponse(BaseModel):
    items: list[SessionRecallSummaryRead]
